src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs

```diff
@@ -249,15 +251,54 @@ private string GetResourceStreamName(CultureInfo culture)
             return resourceStreamName;
         }
 
+        private const string AssemblyElementDelimiter = ", ";
+
+        private string ApplyCultureToAssembly(string assemblyFullName, CultureInfo culture)
+        {
+            var assemblyNameElements = assemblyFullName
+                .Split(new string[] { AssemblyElementDelimiter }, StringSplitOptions.RemoveEmptyEntries);
+            IDictionary<string, string> dict = new Dictionary<string, string>();
+
+            foreach (var assemblyNameElement in assemblyNameElements)
+            {
+                var parts = assemblyNameElement.Split(new char[] { '=' });
+
+                if (parts.Count() == 1)
+                {
+                    parts[0] = parts[0] + ".resources";
+                }
+
+                dict.Add(parts[0], parts.Count() == 2 ? parts[1] : null);
+            }
+
+            // Overwrite any existing culture
+            dict["Culture"] = culture.Name == string.Empty ? "neutral" : culture.Name;
+
+            var elements = dict.Select(kvp => kvp.Value == null ? kvp.Key : $"{kvp.Key}={kvp.Value}");
+
+            return string.Join(AssemblyElementDelimiter, elements);
+        }
+
         private IList<string> GetResourceNamesForCulture(CultureInfo culture)
         {
-            var resourceStreamName = GetResourceStreamName(culture);
+            var assemblyName = ApplyCultureToAssembly(_resourceAssemblyWrapper.FullName, culture);
+
+            Assembly assembly;
+            try
+            {
+                assembly = Assembly.Load(new AssemblyName(assemblyName));
+            }
+            catch (FileNotFoundException)
+            {
+                return null;
+            }
```

Comment:
You are correct, in addition I should be using `assemblyName` in the cache key rather than `_resourceAssemblyWrapper.FullName` so that we cache the correct thing.